docs/docs/migrating-from-v1-to-v2.md

```diff
@@ -0,0 +1,196 @@
+
+### 1. Convert children from function to normal prop (required)
+In v1, the `children` prop passed to layout was a function and needed to be executed. In v2, this is no longer the case.
+
+`layout in v1`
+
+```jsx
+import React from "react"
+
+export default ({ children }) => (
+  <div>
+    {children()}
+  </div>
+)
+```
+
+`layout in v2`
+
+```jsx{5}
+import React from "react"
+
+export default ({ children }) => (
+  <div>
+    {children}
+  </div>
+)
+```
+
+### 2. Move `layout/index.js` to `src/components/layout.js` (optional, but recommended)
+```bash
+git mv src/layouts/index.js src/components/layout.js
+```
+
+### 3. Import and wrap pages with layout component
+Adhering to normal React composition model, you import the layout component and use it to wrap the content of the page.
+
+`src/pages/index.js`
+
+```jsx
+import React from "react"
+import Layout from "../components/layout"
+
+export default () => (
+  <Layout>
+    <div>Hello World</div>
+  </Layout>
+)
+```
+
+Repeat for every page and template that needs this layout.
+
+### 4. Change query to use `StaticQuery`
+<!-- TODO: link StaticQuery text to StaticQuery doc page (if one will be made) -->
+Since layout is no longer special, you now need to make use of v2’s StaticQuery feature.
+
+`Query with layout in v1`
+
+```jsx
+import React, { Fragment } from "react"
+import Helmet from "react-helmet"
+
+export default ({ children, data }) => (
+  <Fragment>
+    <Helmet titleTemplate={`%s | ${data.site.siteMetadata.title}`} defaultTitle={data.site.siteMetadata.title} />
+    <div>
+      {children()}
+    </div>
+  </Fragment>
+)
+
+export const query = graphql`
+  query LayoutQuery {
+    site {
+      siteMetadata {
+        title
+      }
+    }
+  }
+`
+```
+
+`StaticQuery with layout in v2`
+
+```jsx
+import React, { Fragment } from "react"
+import { StaticQuery } from "gatsby"
+
+export default ({ children }) => (
+  <StaticQuery
+    query={graphql`
+      query LayoutQuery {
+        site {
+          siteMetadata {
+            title
+          }
+        }
+      }
+    `}
+    render={data => (
+      <Fragment>
+        <Helmet titleTemplate={`%s | ${data.site.siteMetadata.title}`} defaultTitle={data.site.siteMetadata.title} />
+        <div>
+          {children}
+        </div>
+      </Fragment>
+    )}
+  />
+)
+```
+
+### 5. Pass `history`, `location`, and `match` props to layout
+In v1, layout component had access to `history`, `location`, and `match` props. In v2, only pages have access to these props; pass them to your layout component as needed.
+
+`layout`
+
+```jsx
+import React from "react"
+
+export default ({ children, location }) => (
+  <div>
+    <p>Path is {location.pathname}</p>
+    {children}
+  </div>
+)
+```
+
+`src/pages/index.js`
+
+```jsx
+import React from "react"
+import Layout from "../components/layout.js"
+
+export default props => (
+  <Layout location={props.location}>
+    <div>Hello World</div>
+  </Layout>
+)
+```
+
```

Comment:
There should be a better migration strategy here i think, for sites with lots of pages, this edit will take _forever_ since they have to go into each site and manually wrap them with a layout.

I would suggest demostating a migration path that uses `replaceComponentRenderer` api to automatically wrap every page in the right layout. That or wrap the logic up into a gatsby plugin `gatsby-plugin-layout` or something?

Reply:
> a migration path that uses replaceComponentRenderer api to automatically wrap every page in the right layout

Oh neat, yeah. Have you done this? If you've got an example of using `replaceComponentRenderer` like this, I can write it up into this guide.

> wrap the logic up into a gatsby plugin `gatsby-plugin-layout`

This sounds like a good idea too.

Reply:
We also talked about adding a codemod https://github.com/gatsbyjs/gatsby/issues/5038